File: tests/unit/types.test.ts
import { describe, test, expect } from "vitest";
import type { CodeSymbol, SymbolType, SearchOptions, SearchResult, IndexedFile } from "../../src/types.js";

describe("Type definitions", () => {
	describe("CodeSymbol", () => {
		test("should accept valid CodeSymbol objects", () => {
			const symbol: CodeSymbol = {
				name: "testFunction",
				type: "function",
				file: "/src/test.ts",
				line: 10,
				column: 5,
				context: "function testFunction() {",
			};

			expect(symbol.name).toBe("testFunction");
			expect(symbol.type).toBe("function");
			expect(symbol.file).toBe("/src/test.ts");
			expect(symbol.line).toBe(10);
			expect(symbol.column).toBe(5);
			expect(symbol.context).toBe("function testFunction() {");
		});

		test("should accept CodeSymbol without optional context", () => {
			const symbol: CodeSymbol = {
				name: "testVariable",
				type: "variable",
				file: "/src/test.ts",
				line: 5,
				column: 1,
			};

			expect(symbol.context).toBeUndefined();
		});
	});

	describe("SymbolType", () => {
		test("should include all expected symbol types", () => {
			const validTypes: SymbolType[] = [
				"function",
				"variable", 
				"class",
				"interface",
				"type",
				"enum",
				"constant",
				"method",
				"property",
				"filename",
				"dirname",
			];

			// This test ensures all types are recognized by TypeScript
			validTypes.forEach(type => {
				const symbol: CodeSymbol = {
					name: "test",
					type: type,
					file: "/test.ts",
					line: 1,
					column: 1,
				};
				expect(symbol.type).toBe(type);
			});
		});
	});

	describe("SearchOptions", () => {
		test("should accept empty options object", () => {
			const options: SearchOptions = {};
			expect(options).toEqual({});
		});

		test("should accept all optional properties", () => {
			const options: SearchOptions = {
				includeFiles: false,
				includeDirs: true,
				types: ["function", "class"],
				threshold: 0.5,
				limit: 10,
			};

			expect(options.includeFiles).toBe(false);
			expect(options.includeDirs).toBe(true);
			expect(options.types).toEqual(["function", "class"]);
			expect(options.threshold).toBe(0.5);
			expect(options.limit).toBe(10);
		});

		test("should accept partial options", () => {
			const options1: SearchOptions = { limit: 5 };
			const options2: SearchOptions = { types: ["variable"] };
			const options3: SearchOptions = { threshold: 0.2, includeFiles: false };

			expect(options1.limit).toBe(5);
			expect(options2.types).toEqual(["variable"]);
			expect(options3.threshold).toBe(0.2);
			expect(options3.includeFiles).toBe(false);
		});
	});

	describe("SearchResult", () => {
		test("should structure search results correctly", () => {
			const symbol: CodeSymbol = {
				name: "testSymbol",
				type: "function",
				file: "/test.ts",
				line: 1,
				column: 1,
			};

			const result: SearchResult = {
				symbol: symbol,
				score: 0.25,
			};

			expect(result.symbol).toBe(symbol);
			expect(result.score).toBe(0.25);
		});
	});

	describe("IndexedFile", () => {
		test("should structure indexed file data correctly", () => {
			const symbols: CodeSymbol[] = [
				{
					name: "testFunction",
					type: "function",
					file: "/test.ts",
					line: 1,
					column: 1,
				}
			];

			const indexedFile: IndexedFile = {
				path: "/test.ts",
				symbols: symbols,
				lastModified: Date.now(),
			};

			expect(indexedFile.path).toBe("/test.ts");
			expect(indexedFile.symbols).toBe(symbols);
			expect(typeof indexedFile.lastModified).toBe("number");
		});
	});
});

File: tests/types_test.rs
// TypeScript types.test.ts をRustに移植
// 目標: 8つのテストすべてをパスする

use sfs::types::*;
use std::path::PathBuf;

#[cfg(test)]
mod type_definitions {
    use super::*;

    mod code_symbol {
        use super::*;

        #[test]
        fn should_accept_valid_code_symbol_objects() {
            let symbol = CodeSymbol {
                name: "testFunction".to_string(),
                symbol_type: SymbolType::Function,
                file: "/src/test.ts".into(),
                line: 10,
                column: 5,
                context: Some("function testFunction() {".to_string()),
            };

            assert_eq!(symbol.name, "testFunction");
            assert_eq!(symbol.symbol_type, SymbolType::Function);
            assert_eq!(symbol.file, PathBuf::from("/src/test.ts"));
            assert_eq!(symbol.line, 10);
            assert_eq!(symbol.column, 5);
            assert_eq!(symbol.context, Some("function testFunction() {".to_string()));
        }

        #[test]
        fn should_accept_code_symbol_without_optional_context() {
            let symbol = CodeSymbol {
                name: "testVariable".to_string(),
                symbol_type: SymbolType::Variable,
                file: "/src/test.ts".into(),
                line: 5,
                column: 1,
                context: None,
            };

            assert_eq!(symbol.context, None);
        }
    }

    mod symbol_type {
        use super::*;

        #[test]
        fn should_include_all_expected_symbol_types() {
            let valid_types = vec![
                SymbolType::Function,
                SymbolType::Variable,
                SymbolType::Class,
                SymbolType::Interface,
                SymbolType::Type,
                SymbolType::Enum,
                SymbolType::Constant,
                SymbolType::Method,
                SymbolType::Property,
                SymbolType::Filename,
                SymbolType::Dirname,
            ];

            // TypeScriptテストと同様に、各型でシンボルを作成してテスト
            for symbol_type in valid_types {
                let symbol = CodeSymbol {
                    name: "test".to_string(),
                    symbol_type: symbol_type.clone(),
                    file: "/test.ts".into(),
                    line: 1,
                    column: 1,
                    context: None,
                };
                assert_eq!(symbol.symbol_type, symbol_type);
            }
        }
    }

    mod search_options {
        use super::*;

        #[test]
        fn should_accept_empty_options_object() {
            let options = SearchOptions::default();
            // デフォルト値の検証
            assert_eq!(options, SearchOptions::default());
        }

        #[test]
        fn should_accept_all_optional_properties() {
            let options = SearchOptions {
                include_files: Some(false),
                include_dirs: Some(true),
                types: Some(vec![SymbolType::Function, SymbolType::Class]),
                threshold: Some(0.5),
                limit: Some(10),
            };

            assert_eq!(options.include_files, Some(false));
            assert_eq!(options.include_dirs, Some(true));
            assert_eq!(options.types, Some(vec![SymbolType::Function, SymbolType::Class]));
            assert_eq!(options.threshold, Some(0.5));
            assert_eq!(options.limit, Some(10));
        }

        #[test]
        fn should_accept_partial_options() {
            let options1 = SearchOptions {
                limit: Some(5),
                ..Default::default()
            };
            let options2 = SearchOptions {
                types: Some(vec![SymbolType::Variable]),
                ..Default::default()
            };
            let options3 = SearchOptions {
                threshold: Some(0.2),
                include_files: Some(false),
                ..Default::default()
            };

            assert_eq!(options1.limit, Some(5));
            assert_eq!(options2.types, Some(vec![SymbolType::Variable]));
            assert_eq!(options3.threshold, Some(0.2));
            assert_eq!(options3.include_files, Some(false));
        }
    }

    mod search_result {
        use super::*;

        #[test]
        fn should_structure_search_results_correctly() {
            let symbol = CodeSymbol {
                name: "testSymbol".to_string(),
                symbol_type: SymbolType::Function,
                file: "/test.ts".into(),
                line: 1,
                column: 1,
                context: None,
            };

            let result = SearchResult {
                symbol: symbol.clone(),
                score: 0.25,
            };

            assert_eq!(result.symbol, symbol);
            assert_eq!(result.score, 0.25);
        }
    }

    mod indexed_file {
        use super::*;

        #[test]
        fn should_structure_indexed_file_data_correctly() {
            let symbols = vec![CodeSymbol {
                name: "testFunction".to_string(),
                symbol_type: SymbolType::Function,
                file: "/test.ts".into(),
                line: 1,
                column: 1,
                context: None,
            }];

            let now = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64;

            let indexed_file = IndexedFile {
                path: "/test.ts".into(),
                symbols: symbols.clone(),
                last_modified: now,
            };

            assert_eq!(indexed_file.path, PathBuf::from("/test.ts"));
            assert_eq!(indexed_file.symbols, symbols);
            assert!(indexed_file.last_modified > 0);
        }
    }
}